File: ECodeforces-150/B.cpp
// https://codeforces.com/contest/1841/problem/B
#include <iostream>
#include <vector>
using namespace std;

int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        int q;
        cin >> q;
        vector<int> queries(q);
        for (int i = 0; i < q; i++)
        {
            cin >> queries[i];
        }
        vector<int> res;

        int count = 0; // queries[j] > queries[j+1] count

        for (int i = 0; i < q; i++)
        {
            int new_count = count + (res.size() && res.back() > queries[i]);

            if (new_count == 0 || (new_count == 1 && queries[i] <= res[0]))
            {
                res.push_back(queries[i]);
                count = new_count; // count = 0 or 1
                cout << '1';
               
            }
            else
            {
                cout << '0';
            }
        }
        cout << endl;
    }
}

File: ECodeforces-150/B.rs
use std::io::{self, BufRead};
//https://codeforces.com/contest/1841/problem/B
fn main() {
    let stdin = io::stdin();
    let mut buffer = String::new();
    stdin.read_line(&mut buffer).unwrap();
    let t: usize = buffer.trim().parse().unwrap();

    for _ in 0..t {
        buffer.clear();
        stdin.read_line(&mut buffer).unwrap();
        let q: usize = buffer.trim().parse().unwrap();

        buffer.clear();
        stdin.read_line(&mut buffer).unwrap();
        let queries: Vec<i32> = buffer
            .split_whitespace()
            .map(|x| x.parse().unwrap())
            .collect();

        let mut res = Vec::new();
        let mut count = 0;

        for i in 0..q {
            let new_count = count + (!res.is_empty() && res.last().unwrap() > &queries[i]) as i32;

            if new_count == 0 || (new_count == 1 && queries[i] <= res[0]) {
                res.push(queries[i]);
                count = new_count;
                print!("1");
            } else {
                print!("0");
            }
        }
        println!();
    }
}
